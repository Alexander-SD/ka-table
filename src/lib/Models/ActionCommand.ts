export class ActionCommand {
  private rejectedFlag: boolean = false;
  get rejected() {
    return this.rejectedFlag;
  }
  public reject(): void {
    this.rejectedFlag = true;
  }
}
