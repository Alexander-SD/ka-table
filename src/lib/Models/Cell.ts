/**
 * Describes the position of cell in table
 */
export class Cell {
  public columnKey!: string;
  public rowKeyValue!: any;
}
