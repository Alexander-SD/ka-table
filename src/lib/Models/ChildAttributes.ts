import { ITableAllProps } from '../';
import { ICellContentProps } from '../Components/CellContent/CellContent';
import { IDataRowProps } from '../Components/DataRowContent/DataRowContent';
import { ChildAttributesItem } from '../types';

export class ChildAttributes {
  public cell?: ChildAttributesItem<ICellContentProps>;
  public table?: ChildAttributesItem<ITableAllProps>;
  public dataRow?: ChildAttributesItem<IDataRowProps>;
}
