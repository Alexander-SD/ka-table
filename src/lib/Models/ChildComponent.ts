import { PropsWithChildren } from 'react';

import { ChildAttributesItem } from '../types';

export class ChildComponent<TProps> {
  public elementAttributes?: ChildAttributesItem<TProps>;
  public propsAreEqual?: (prevProps: TProps, nextProps: TProps) => boolean;
  public content?: (props: PropsWithChildren<TProps>) => any
}