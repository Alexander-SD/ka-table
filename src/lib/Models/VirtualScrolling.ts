export class VirtualScrolling {
  public itemHeight!: (data: any) => number | number;
  public scrollPosition?: number;
  public visibleItemsCount?: number;
}
