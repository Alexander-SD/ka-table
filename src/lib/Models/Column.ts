import { FilterCondition } from '../../../dist/Models/FilterCondition';
import defaultOptions from '../defaultOptions';
import { DataType, SortDirection } from '../enums';
import {
  CellFunc, EditorFunc, FormatFunc, HeaderCellFunc, SearchFunc, ValidationFunc,
} from '../types';

/**
 * Describes column of table its look and behaviour
 */
export class Column {
  public cell?: CellFunc;
  public dataType?: DataType = defaultOptions.columnDataType;
  public editor?: EditorFunc;
  public filterRowCell?: EditorFunc;
  public filterRowValue?: any;
  public filterRowOperator?: any;
  public field?: string;
  public format?: FormatFunc;
  public headCell?: HeaderCellFunc;
  public isEditable?: boolean;
  public key!: string;
  public search?: SearchFunc;
  public sortDirection?: SortDirection;
  public style?: React.CSSProperties;
  public title?: string;
  public validation?: ValidationFunc;
}
