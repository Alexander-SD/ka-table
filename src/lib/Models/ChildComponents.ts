import { ITableProps } from '../';
import { ICellProps } from '../Components/CellComponent/CellComponent';
import { ICellContentProps } from '../Components/CellContent/CellContent';
import { ICellEditorProps, IFilterRowEditorProps } from '../Components/CellEditor/CellEditor';
import { IDataRowProps } from '../Components/DataRowContent/DataRowContent';
import { IGroupRowProps } from '../Components/GroupRowContent/GroupRowContent';
import { IHeadCellProps } from '../Components/HeadCell/HeadCell';
import { ChildComponent } from './ChildComponent';

export class ChildComponents {
  public cell?: ChildComponent<ICellProps>;
  public editor?: ChildComponent<ICellEditorProps>;
  public filterRowCell?: ChildComponent<IFilterRowEditorProps>;
  public groupCell?: ChildComponent<IGroupRowProps>;
  public headCell?: ChildComponent<IHeadCellProps>;
  public cellText?: ChildComponent<ICellContentProps>;
  public table?: ChildComponent<ITableProps>;
  public dataRow?: ChildComponent<IDataRowProps>;
  public detailsRow?: ChildComponent<IDataRowProps>;
}
