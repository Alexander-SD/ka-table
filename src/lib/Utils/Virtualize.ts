import { VirtualScrolling } from '../Models/VirtualScrolling';

export const getVirtualized = (virtualScrolling: VirtualScrolling, data: any[]) => {
  const virtualizedData: any[] = [];
  const { scrollPosition = 0, visibleItemsCount = 40 } = virtualScrolling;
  let beginHeight = 0;
  let endHeight = 0;
  data.reduce((acc, value) => {
    const itemHeight = virtualScrolling.itemHeight(value);
    if (acc > scrollPosition - itemHeight) {
      if (virtualizedData.length < visibleItemsCount) {
        virtualizedData.push(value);
      } else {
        endHeight += itemHeight;
      }
    } else {
      beginHeight = acc - itemHeight;
    }
    return acc + itemHeight;
  }, 0);

  return {
    beginHeight,
    endHeight,
    virtualizedData,
  };
};
