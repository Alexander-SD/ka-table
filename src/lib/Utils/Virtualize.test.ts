import { VirtualScrolling } from '../Models/VirtualScrolling';
import { getVirtualized } from './Virtualize';

describe('Virtualize', () => {
  const data: any[] = new Array(200).fill(1);
  describe('getVirtualized', () => {
    it('scrollPosition 0', () => {
      const virtualScrolling: VirtualScrolling = {
        itemHeight: () => 10,
        scrollPosition: 0,
        visibleItemsCount: 4,
      };
      const result = getVirtualized(virtualScrolling, data);

      expect(result).toMatchSnapshot();
    });

    it('scrollPosition 100', () => {
      const virtualScrolling: VirtualScrolling = {
        itemHeight: () => 10,
        scrollPosition: 101,
        visibleItemsCount: 4,
      };
      const result = getVirtualized(virtualScrolling, data);

      expect(result).toMatchSnapshot();
    });

    it('scrollPosition 2000', () => {
      const virtualScrolling: VirtualScrolling = {
        itemHeight: () => 10,
        scrollPosition: 2000,
        visibleItemsCount: 4,
      };
      const result = getVirtualized(virtualScrolling, data);

      expect(result).toMatchSnapshot();
    });

    it('itemHeight 40', () => {
      const virtualScrolling: VirtualScrolling = {
        itemHeight: () => 40,
        scrollPosition: 0,
        visibleItemsCount: 4,
      };
      const result = getVirtualized(virtualScrolling, data);

      expect(result).toMatchSnapshot();
    });
  });
});
