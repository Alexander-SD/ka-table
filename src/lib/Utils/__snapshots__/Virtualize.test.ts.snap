// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Virtualize getVirtualized itemHeight 40 1`] = `
Object {
  "beginHeight": 0,
  "endHeight": 7840,
  "virtualizedData": Array [
    1,
    1,
    1,
    1,
  ],
}
`;

exports[`Virtualize getVirtualized scrollPosition 0 1`] = `
Object {
  "beginHeight": 0,
  "endHeight": 1960,
  "virtualizedData": Array [
    1,
    1,
    1,
    1,
  ],
}
`;

exports[`Virtualize getVirtualized scrollPosition 100 1`] = `
Object {
  "beginHeight": 80,
  "endHeight": 1860,
  "virtualizedData": Array [
    1,
    1,
    1,
    1,
  ],
}
`;

exports[`Virtualize getVirtualized scrollPosition 2000 1`] = `
Object {
  "beginHeight": 1980,
  "endHeight": 0,
  "virtualizedData": Array [],
}
`;
